src/Poisson3D.hpp

```diff
@@ -0,0 +1,303 @@
+#ifndef POISSON_3D_HPP
+#define POISSON_3D_HPP
+
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_in.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/tria.h>
+
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/solver_gmres.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_vector.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+
+#include <deal.II/lac/eigen.h> //For power method
+#include <deal.II/lac/vector_memory.h> // Power method
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+#include <vector>
+#include <cmath> // For pow and sqrt functions
+#include <random>
+
+#include <unistd.h> // For command line arguments
+
+using namespace dealii;
+
+/**
+ * Class managing the differential problem.
+ */
+
+class Poisson3D
+{
+public:
+  // Physical dimension (1D, 2D, 3D)
+  static constexpr unsigned int dim = 3;
+
+  // Diffusion coefficient.
+  // In deal.ii, functions are implemented by deriving the dealii::Function
+  // class, which provides an interface for the computation of function values
+  // and their derivatives.
+
+  template <int dim>
+  class DiffusionCoefficient : public Function<dim>
+  {
+  public:
+      // Define a structure to hold sphere data
+      struct Sphere
+      {
+          Point<dim> center;
+          double radius;
+      };
+  
+      // Constructor
+      DiffusionCoefficient(const std::vector<Sphere>& spheres, double p)
+      : spheres(spheres), p_value(p) {}
+
+      DiffusionCoefficient(){}
+
+      // Evaluation
+      virtual double value(const Point<dim>& p, const unsigned int /*component*/ = 0) const override
+      {
+          for (const auto& sphere : spheres)
+          {
+              if (is_point_inside_sphere(p, sphere))
+                  return std::pow(10.0, p_value);
+          }
+          return 1.0;
+      }
+
+  private:
+      std::vector<Sphere> spheres;
+      double p_value;
```

Comment:
If data members are not supposed to change over time, as is the case here, they should be declared as const:
```suggestion
      const std::vector<Sphere> spheres;
      const double p_value;
```